amp-pwa-reader/src/js/Article.js

```diff
@@ -0,0 +1,250 @@
+
+    // resize card to image ratio
+    card.style.height = (innerWidth * this.card.imageData.ratio) + 'px';
+    card.style.opacity = '0';
+    card.style.margin = '0';
+
+    this.clonedCard = card;
+    return card;
+
+  }
+
+  generateCard() {
+
+    let articleData = shadowReader.backend.getArticleData();
+    let card = new Card(articleData, /*headless*/true).elem;
+
+    // resize card to image ratio
+    card.style.height = (innerWidth * articleData.imageRatio) + 'px';
+    card.style.margin = '0';
+
+    this.clonedCard = card;
+    return card;
+
+  }
+
+  getAnimationSpeed() {
+    if (this._animationSpeed)
+      return this._animationSpeed;
+
+    let htmlStyles = window.getComputedStyle(document.querySelector("html"));
+    this._animationSpeed = parseFloat(htmlStyles.getPropertyValue("--animation-speed")) * 1000;
+    return this._animationSpeed;
+  }
+
+  animateIn() {
+
+    // No animation if there's no card to animate from
+    if (!this.card) {
+      return new Promise(resolve => resolve());
+    }
+
+    let offset = (innerWidth * this.card.imageData.ratio) / 2;
+    this.container.style.transform = 'translateY(' + scrollY + 'px)';
+    this.container.style.opacity = 1;
+
+    return new Promise(resolve => {
+      this.container.animate([
+        { opacity: 0, transform: 'translateY(' + (offset + scrollY) + 'px)' },
+        { opacity: 1, transform: 'translateY(' + scrollY + 'px)' }
+      ], { duration: this.getAnimationSpeed(), easing: 'ease-out' }).onfinish = resolve;
+    });
+
+  }
+
+  animateOut() {
+
+    // No animation if there's no card to animate from
+    if (!this.card) {
+        return new Promise(resolve => resolve());
+    }
+
+    let offset = (innerWidth * this.card.imageData.ratio) / 2;
+    this.container.style.transform = 'translateY(' + (offset + scrollY) + 'px)';
+
+    return new Promise(resolve => {
+      return this.container.animate([
+        { opacity: 1, transform: 'translateY(' + (scrollY) + 'px)' },
+        { opacity: 0, transform: 'translateY(' + (offset + scrollY) + 'px)' }
+      ], { duration: this.getAnimationSpeed(), easing: 'ease-out' }).onfinish = resolve;
+    });
+
+  }
+
+  render() {
+
+    // Create an empty container for the AMP page
+    this.container = this.createShadowRoot();
+
+    // Tell Shadow AMP to initialize the AMP page in prerender-mode
+    this.ampDoc = AMP.attachShadowDoc(this.container, this.doc, this.url);
+    this.ampDoc.setVisibilityState('prerender');
+
+    return this.ampDoc.ampdoc.whenReady();
+
+  }
+
+  show(replaceHistoryState) {
+
+    // We need to clone the featured image into the Shadow DOM so it scrolls
+    // along. There are cases were we don't have a linked card from the list
+    // view (e.g. we load directly into the article), in which case we need to
+    // generate a new one.
+    var card = this.card ? this.cloneCard() : this.generateCard();
+    this.ampDoc.ampdoc.getBody().prepend(card);
+
+    return new Promise(resolve => {
```

Comment:
Can this simplified with the below?

     return this.animateIn().then(() => {

         // Hide the original card, show the cloned one (this also animates)
         if (this.card) {
           card.style.opacity = '1';
         }
 
         // add class to html element for to contain the scroll, and transform
         // the hamburger into a 'back' button.
         document.documentElement.classList.add('sr-article-shown');
 
         this.takeoverScroll();
 
         // Set the visibility state of the AMP doc to visible
         this.ampDoc.setVisibilityState('visible');
 
         // Finally, add new history entry
         // Note: We're doing this deliberately late due to an AMP
         // Bug that overrides the history state object early on
         shadowReader.nav.setOpenArticle(this, replaceHistoryState);
         return true;
       });